src/visualization/utils/colorMappingUtils.ts

```diff
@@ -0,0 +1,130 @@
+import {XYViewProperties} from 'src/types'
+
+/**
+ * Evaluates deeper equality for two map objects based on their key:value pair
+ * @param map1
+ * @param map2
+ */
+const areMappingsSame = (map1, map2) => {
+  if (!map1 || !map2) {
+    return false
+  }
+
+  // Create arrays of property names
+  const aProps = Object.getOwnPropertyNames(map1)
+  const bProps = Object.getOwnPropertyNames(map2)
+
+  // If number of properties is different,
+  // objects are not equivalent
+  if (aProps.length != bProps.length) {
+    return false
+  }
+
+  for (let i = 0; i < aProps.length; i++) {
+    const propName = aProps[i]
+
+    // If values of same property are not equal,
+    // objects are not equivalent
+    if (map1[propName] !== map2[propName]) {
+      return false
+    }
+  }
+
+  return true
+}
+
+/**
+ * Returns the colorMapping objects for UI, based on the state of the current view it will return the existing or generate a new one and request trigger saving to IDPE
+ * @param columnGroupMap
+ * @param properties
+ * @returns an object shaped : {colorMappingForIDPE?, colorMappingForGiraffe?, needsToSaveToIDPE}
+ */
+
+export const getColorMappingObjects = (
+  columnGroupMap,
+  properties: XYViewProperties
+) => {
+  const seriesToColorIndexMap = generateSeriesToColorIndexMap(
+    columnGroupMap,
+    properties
+  )
+  let needsToSaveToIDPE = false
+
+  // if the mappings from the IDPE and the *required* one's for the current view are the same, we don't need to generate new mappings
+  if (areMappingsSame(properties.colorMapping, seriesToColorIndexMap)) {
+    const columnKeys = columnGroupMap.columnKeys
+    const mappings = {...columnGroupMap}
+
+    mappings.mappings.forEach(graphLine => {
+      const seriesID = getSeriesId(graphLine, columnKeys)
+
+      const colors = properties.colors
+
+      // this is needed for giraffe
+      graphLine.color = colors[properties.colorMapping[seriesID]].hex
+    })
+
+    needsToSaveToIDPE = false
+
+    return {
+      colorMappingForGiraffe: {columnKeys, ...mappings},
+      needsToSaveToIDPE,
+    }
+  } else {
+    const columnKeys = columnGroupMap.columnKeys
+    const mappings = {...columnGroupMap}
+
+    mappings.mappings.forEach(graphLine => {
+      const seriesID = getSeriesId(graphLine, columnKeys)
+
+      const colors = properties.colors
+
+      // this is needed for giraffe
+      graphLine.color = colors[seriesToColorIndexMap[seriesID]].hex
+    })
+
+    const newColorMappingForGiraffe = {
+      ...mappings,
+      seriesToColorIndexMap,
+    }
+    needsToSaveToIDPE = true
+
+    return {
+      colorMappingForIDPE: seriesToColorIndexMap,
+      colorMappingForGiraffe: newColorMappingForGiraffe,
+      needsToSaveToIDPE,
+    }
+  }
+}
+
```

Comment:
I'd like to see a comment here showing how a group of column keys gets turned into a series id. It would be really illustrative and not need to be very long or detailed.

Reply:
✅